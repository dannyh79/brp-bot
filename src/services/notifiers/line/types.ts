import * as line from '@line/bot-sdk';

export type LineNotifierArg = {
  to: string;
  channelAccessToken: string;
};

export type LineMessage = line.messagingApi.Message;

export type LinePushMessageRequest = line.messagingApi.PushMessageRequest;
