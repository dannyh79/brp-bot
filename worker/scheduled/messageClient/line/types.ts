import * as line from '@line/bot-sdk';

export type LineMessagingApiClientArg = {
  channelAccessToken: string;
};

export type LineFlexComponent = line.messagingApi.FlexComponent;

export type LineMessage = line.messagingApi.Message;

export type LineMessageArg = {
  to: string;
  messages: LineMessage[];
};
