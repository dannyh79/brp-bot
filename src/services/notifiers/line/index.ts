export * from './singleNotifier';
export * from './multiNotifier';
export * from './types';
