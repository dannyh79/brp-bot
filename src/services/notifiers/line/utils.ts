import { GetPlanOutput } from '@/readingPlans';
import { LineMessage } from './types';

const formatDateString = (str: string): { date: string; dayOfWeek: string } => {
  const date = new Date(str);
  if (isNaN(date.getTime())) {
    throw new Error(`Invalid date string: ${str}`);
  }

  const daysOfWeek = ['星期日', '星期一', '星期二', '星期三', '星期四', '星期五', '星期六'];
  const dayOfWeek = daysOfWeek[date.getDay()];

  const month = date.getMonth() + 1; // Months are 0-based
  const day = date.getDate();
  return {
    date: `${month}/${day}`,
    dayOfWeek,
  };
};

export const toBubbleMessage = (arg: GetPlanOutput): LineMessage => {
  const { date: dateFromData, praise, repentence, devotional, prayer } = arg;
  const { date, dayOfWeek } = formatDateString(dateFromData);
  const [personalPrayer, ...mainPrayer] = prayer.split('\n');

  return {
    type: 'flex',
    altText: `Bible Reading Plan for ${dateFromData}`,
    contents: {
      type: 'bubble',
      size: 'giga',
      header: {
        type: 'box',
        layout: 'horizontal',
        contents: [
          {
            type: 'box',
            layout: 'horizontal',
            contents: [
              {
                type: 'text',
                text: '好好靈修',
                color: '#1D292E',
                size: 'xl',
                weight: 'bold',
                flex: 2,
                align: 'end',
              },
              {
                type: 'text',
                text: 'X',
                size: 'md',
                align: 'center',
              },
              {
                type: 'box',
                layout: 'vertical',
                contents: [
                  {
                    type: 'text',
                    text: 'DAILY',
                    weight: 'bold',
                  },
                  {
                    type: 'text',
                    text: 'DEVOTION',
                    weight: 'bold',
                  },
                ],
                flex: 5,
              },
            ],
            alignItems: 'center',
          },
        ],
        paddingAll: '20px',
        backgroundColor: '#FFCC32',
        spacing: 'md',
      },
      body: {
        type: 'box',
        layout: 'vertical',
        contents: [
          {
            type: 'text',
            text: `${date} ${dayOfWeek}`,
            offsetBottom: 'md',
          },
          {
            type: 'box',
            layout: 'vertical',
            contents: [
              {
                type: 'box',
                layout: 'horizontal',
                contents: [
                  {
                    type: 'box',
                    layout: 'vertical',
                    contents: [
                      {
                        type: 'text',
                        text: '讚',
                        size: 'lg',
                      },
                      {
                        type: 'text',
                        text: '美',
                        size: 'lg',
                      },
                    ],
                    spacing: 'xl',
                    justifyContent: 'center',
                    alignItems: 'center',
                    flex: 1,
                  },
                  {
                    type: 'text',
                    text: `${praise.content}\n(${praise.scope})`,
                    flex: 11,
                    size: 'xs',
                    color: '#1C1C1D',
                    wrap: true,
                  },
                ],
                paddingStart: 'md',
                paddingEnd: 'md',
                spacing: 'lg',
              },
              {
                type: 'box',
                layout: 'horizontal',
                contents: [],
                backgroundColor: '#FFFFFFAA',
                borderWidth: 'light',
                width: '95%',
                offsetStart: '5%',
                offsetEnd: '5%',
              },
            ],
            spacing: 'lg',
            cornerRadius: 'md',
            backgroundColor: '#C7CADA',
            paddingTop: 'lg',
            paddingBottom: 'md',
          },
          {
            type: 'box',
            layout: 'vertical',
            contents: [
              {
                type: 'box',
                layout: 'horizontal',
                contents: [
                  {
                    type: 'box',
                    layout: 'vertical',
                    contents: [
                      {
                        type: 'text',
                        text: '悔',
                        size: 'lg',
                      },
                      {
                        type: 'text',
                        text: '改',
                        size: 'lg',
                      },
                    ],
                    flex: 1,
                    spacing: 'xl',
                    justifyContent: 'center',
                    alignItems: 'center',
                  },
                  {
                    type: 'text',
                    text: repentence,
                    flex: 11,
                    size: 'xs',
                    color: '#1C1C1D',
                    wrap: true,
                  },
                ],
                spacing: 'lg',
                paddingStart: 'md',
                paddingEnd: 'md',
              },
              {
                type: 'box',
                layout: 'horizontal',
                contents: [],
                backgroundColor: '#C7CADAAA',
                borderWidth: 'light',
                width: '95%',
                offsetStart: '5%',
                offsetEnd: '5%',
              },
            ],
            spacing: 'lg',
            paddingBottom: 'md',
            paddingTop: 'lg',
            backgroundColor: '#FFFFFF',
          },
          {
            type: 'box',
            layout: 'vertical',
            contents: [
              {
                type: 'box',
                layout: 'horizontal',
                contents: [
                  {
                    type: 'box',
                    layout: 'vertical',
                    contents: [
                      {
                        type: 'text',
                        text: '反',
                        size: 'lg',
                      },
                      {
                        type: 'text',
                        text: '思',
                        size: 'lg',
                        align: 'start',
                      },
                    ],
                    flex: 1,
                    spacing: 'xl',
                    justifyContent: 'center',
                    alignItems: 'center',
                  },
                  {
                    type: 'box',
                    layout: 'vertical',
                    contents: [
                      {
                        type: 'text',
                        text: devotional.scope,
                        size: 'lg',
                        weight: 'bold',
                        wrap: true,
                      },
                      {
                        type: 'text',
                        text: devotional.content.join('\n'),
                        size: 'xs',
                        color: '#1C1C1D',
                        wrap: true,
                        flex: 9,
                      },
                    ],
                    flex: 11,
                    paddingTop: 'none',
                    spacing: 'md',
                  },
                ],
                paddingStart: 'md',
                paddingEnd: 'md',
                spacing: 'lg',
              },
              {
                type: 'box',
                layout: 'horizontal',
                contents: [],
                backgroundColor: '#FFFFFFAA',
                borderWidth: 'light',
                width: '95%',
                offsetStart: '5%',
                offsetEnd: '5%',
              },
            ],
            spacing: 'lg',
            cornerRadius: 'md',
            backgroundColor: '#C7CADA',
            paddingTop: 'lg',
            paddingBottom: 'md',
          },
          {
            type: 'box',
            layout: 'vertical',
            contents: [
              {
                type: 'box',
                layout: 'horizontal',
                contents: [
                  {
                    type: 'box',
                    layout: 'vertical',
                    contents: [
                      {
                        type: 'text',
                        text: '祈',
                        size: 'lg',
                      },
                      {
                        type: 'text',
                        text: '求',
                        size: 'lg',
                      },
                    ],
                    justifyContent: 'center',
                    alignItems: 'center',
                    flex: 1,
                    spacing: 'xl',
                  },
                  {
                    type: 'box',
                    layout: 'vertical',
                    flex: 11,
                    contents: [
                      {
                        type: 'text',
                        text: personalPrayer,
                        size: 'xs',
                        color: '#1C1C1D',
                        wrap: true,
                      },
                      {
                        type: 'box',
                        layout: 'horizontal',
                        contents: [
                          {
                            type: 'box',
                            layout: 'horizontal',
                            contents: [
                              {
                                type: 'box',
                                layout: 'vertical',
                                contents: [],
                                borderWidth: 'light',
                                borderColor: '#E2E2E299',
                                width: '1px',
                              },
                            ],
                            flex: 1,
                          },
                          {
                            type: 'text',
                            text: mainPrayer.join('\n'),
                            size: 'xs',
                            color: '#1C1C1DAA',
                            wrap: true,
                            flex: 24,
                          },
                        ],
                      },
                    ],
                    spacing: 'md',
                  },
                ],
                paddingStart: 'md',
                paddingEnd: 'md',
                spacing: 'sm',
              },
              {
                type: 'box',
                layout: 'horizontal',
                contents: [],
                backgroundColor: '#C7CADAAA',
                borderWidth: 'light',
                width: '95%',
                offsetStart: '5%',
                offsetEnd: '5%',
              },
            ],
            spacing: 'lg',
            backgroundColor: '#FFFFFF',
            paddingTop: 'lg',
            paddingBottom: 'md',
          },
        ],
      },
    },
  };
};
