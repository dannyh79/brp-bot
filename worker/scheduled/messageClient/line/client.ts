import { MessageClient } from '../types';
import { LineMessage, LineMessageArg, LineMessagingApiClientArg } from './types';

/** Implement line's MessagingApiClient, for @line/bot-sdk has worker incompatible dep "axios". */
class LineMessagingApiClient implements MessageClient<LineMessageArg, LineMessage[]> {
  private channelAccessToken: string;
  /** LINE Messaging API base URL. */
  private baseUrl: string = 'https://api.line.me/v2/bot';

  constructor({ channelAccessToken }: LineMessagingApiClientArg) {
    this.channelAccessToken = channelAccessToken;
  }

  async pushMessage({ to, messages }: LineMessageArg): Promise<LineMessage[]> {
    const url = `${this.baseUrl}/message/push`;
    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.channelAccessToken}`,
      },
      body: JSON.stringify({
        to,
        messages,
      }),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Failed to push message: ${response.status} ${errorText}`);
    }

    return response.json();
  }
}

export const createClient = (channelAccessToken: string) =>
  new LineMessagingApiClient({ channelAccessToken });

export default createClient;
