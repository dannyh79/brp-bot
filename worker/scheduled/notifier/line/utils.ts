import * as line from '@line/bot-sdk';
import { GetPlanOutput } from '@/usecases/getPlan';
import { LineMessage } from './types';

export const toBubbleMessage = (arg: GetPlanOutput): LineMessage => {
  const { date, scope, content } = arg;

  const toParagraph = (text: string): line.messagingApi.FlexComponent => ({
    type: 'text',
    text,
    wrap: true,
    color: '#666666',
    size: 'sm',
    flex: 5,
  });

  return {
    type: 'flex',
    altText: `Bible Reading Plan for ${date}`,
    contents: {
      type: 'bubble',
      body: {
        type: 'box',
        layout: 'vertical',
        contents: [
          {
            type: 'text',
            text: date,
            weight: 'bold',
            size: 'xl',
          },
          {
            type: 'box',
            layout: 'vertical',
            margin: 'lg',
            spacing: 'sm',
            contents: [
              {
                type: 'box',
                layout: 'baseline',
                spacing: 'sm',
                contents: [
                  {
                    type: 'text',
                    text: '範圍',
                    color: '#aaaaaa',
                    size: 'sm',
                    flex: 1,
                  },
                  {
                    type: 'text',
                    text: scope,
                    wrap: true,
                    color: '#666666',
                    size: 'sm',
                    flex: 5,
                  },
                ],
              },
              {
                type: 'box',
                layout: 'vertical',
                spacing: 'sm',
                contents: content.map(toParagraph),
              },
            ],
          },
        ],
      },
    },
  };
};
