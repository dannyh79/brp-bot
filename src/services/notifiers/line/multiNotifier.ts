import { GetPlanOutput } from '@/readingPlans';
import { Notifier } from '../types';
import { LineMultiNotifierArg, LinePushMessageRequest } from './types';
import * as utils from './utils';

/**
 * Implement line's MessagingApiClient, for "@line/bot-sdk" has worker incompatible dep "axios".
 *
 * @see {@link https://developers.line.biz/en/reference/messaging-api/#send-push-message} for API doc
 */
export class LineMultiNotifier implements Notifier<string[], GetPlanOutput> {
  private channelAccessToken: string;
  /** LINE Messaging API base URL. */
  private baseUrl: string = 'https://api.line.me/v2/bot';

  constructor({ channelAccessToken }: LineMultiNotifierArg) {
    this.channelAccessToken = channelAccessToken;
  }

  async pushMessage(recipients: string[], message: GetPlanOutput) {
    const reqs = recipients.map(async (to) => {
      const url = `${this.baseUrl}/message/push`;
      const payload: LinePushMessageRequest = {
        to,
        messages: [utils.toBubbleMessage(message)],
      };

      const response = await fetch(url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Authorization': `Bearer ${this.channelAccessToken}`,
        },
        body: JSON.stringify(payload),
      });

      if (!response.ok) {
        const errorText = await response.text();
        throw new Error(`Failed to push message: ${response.status} ${errorText}`);
      }

      return response.json();
    });

    return Promise.allSettled(reqs);
  }
}

export default LineMultiNotifier;
