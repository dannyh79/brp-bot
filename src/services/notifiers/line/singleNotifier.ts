import { GetPlanOutput } from '@/readingPlans';
import { Notifier } from '../types';
import { LineMessage, LinePushMessageRequest, LineSingleNotifierArg } from './types';
import * as utils from './utils';

/** Implement line's MessagingApiClient, for @line/bot-sdk has worker incompatible dep "axios". */
export class LineSingleNotifier implements Notifier<GetPlanOutput, LineMessage[]> {
  private channelAccessToken: string;
  private to: string;
  /** LINE Messaging API base URL. */
  private baseUrl: string = 'https://api.line.me/v2/bot';

  constructor({ channelAccessToken, to }: LineSingleNotifierArg) {
    this.channelAccessToken = channelAccessToken;
    this.to = to;
  }

  async pushMessage(message: GetPlanOutput): Promise<LineMessage[]> {
    const url = `${this.baseUrl}/message/push`;
    const payload: LinePushMessageRequest = {
      to: this.to,
      messages: [utils.toBubbleMessage(message)],
    };

    const response = await fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Authorization': `Bearer ${this.channelAccessToken}`,
      },
      body: JSON.stringify(payload),
    });

    if (!response.ok) {
      const errorText = await response.text();
      throw new Error(`Failed to push message: ${response.status} ${errorText}`);
    }

    return response.json();
  }
}

export default LineSingleNotifier;
