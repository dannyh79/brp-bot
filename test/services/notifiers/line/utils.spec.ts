import { toBubbleMessage } from '@/services/notifiers/line/utils';

describe('toBubbleMessage()', () => {
  const data = {
    date: '2025-01-01',
    praise: {
      scope: '歷代志上 16:34 CCB',
      content: '你們要稱謝耶和華, 因為祂是美善的, 祂的慈愛永遠長存!',
    },
    repentence:
      '上帝啊，求你按你的慈愛恩待我！\n按你極大的憐憫除去我 ___ 的過犯！\n求你洗淨我的罪過，清除我的罪惡。求你讓我重新享受蒙你拯救的喜樂，賜我一個樂意順服你的心靈。',
    devotional: {
      scope: '出埃及記 第 8 章',
      content: ['1. 你覺得神透過今天的經文對你說什麼呢?', '2. 有什麼你可以做出的行動或改變呢?'],
    },
    prayer:
      '”神啊！我將我的，___ ， ___ ， ___ 交給祢，我相信祢在這些事上掌權。\n『我們在天上的父：願人都尊你的名為聖。願你的國降臨；願你的旨意行在地上，如同行在天上。我們日用的飲食，今日賜給我們。免我們的債，如同我們免了人的債。不叫我們陷入試探；救我們脫離那惡者。因為國度、權柄、榮耀，全是你的，直到永遠。阿們！』”',
  };

  it('returns LINE bubble message payload', () => {
    expect(toBubbleMessage(data)).toMatchSnapshot();
  });
});
