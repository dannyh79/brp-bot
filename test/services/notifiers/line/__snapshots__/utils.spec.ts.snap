// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`toBubbleMessage() > returns LINE bubble message payload 1`] = `
{
  "altText": "Bible Reading Plan for 2025-01-01",
  "contents": {
    "body": {
      "contents": [
        {
          "offsetBottom": "md",
          "text": "1/1 星期三",
          "type": "text",
        },
        {
          "backgroundColor": "#C7CADA",
          "contents": [
            {
              "contents": [
                {
                  "alignItems": "center",
                  "contents": [
                    {
                      "size": "lg",
                      "text": "讚",
                      "type": "text",
                    },
                    {
                      "size": "lg",
                      "text": "美",
                      "type": "text",
                    },
                  ],
                  "flex": 1,
                  "justifyContent": "center",
                  "layout": "vertical",
                  "spacing": "xl",
                  "type": "box",
                },
                {
                  "color": "#1C1C1D",
                  "flex": 11,
                  "size": "xs",
                  "text": "你們要稱謝耶和華, 因為祂是美善的, 祂的慈愛永遠長存!
(歷代志上 16:34 CCB)",
                  "type": "text",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingEnd": "md",
              "paddingStart": "md",
              "spacing": "lg",
              "type": "box",
            },
            {
              "backgroundColor": "#FFFFFFAA",
              "borderWidth": "light",
              "contents": [],
              "layout": "horizontal",
              "offsetEnd": "5%",
              "offsetStart": "5%",
              "type": "box",
              "width": "95%",
            },
          ],
          "cornerRadius": "md",
          "layout": "vertical",
          "paddingBottom": "md",
          "paddingTop": "lg",
          "spacing": "lg",
          "type": "box",
        },
        {
          "backgroundColor": "#FFFFFF",
          "contents": [
            {
              "contents": [
                {
                  "alignItems": "center",
                  "contents": [
                    {
                      "size": "lg",
                      "text": "悔",
                      "type": "text",
                    },
                    {
                      "size": "lg",
                      "text": "改",
                      "type": "text",
                    },
                  ],
                  "flex": 1,
                  "justifyContent": "center",
                  "layout": "vertical",
                  "spacing": "xl",
                  "type": "box",
                },
                {
                  "color": "#1C1C1D",
                  "flex": 11,
                  "size": "xs",
                  "text": "上帝啊，求你按你的慈愛恩待我！
按你極大的憐憫除去我 ___ 的過犯！
求你洗淨我的罪過，清除我的罪惡。求你讓我重新享受蒙你拯救的喜樂，賜我一個樂意順服你的心靈。",
                  "type": "text",
                  "wrap": true,
                },
              ],
              "layout": "horizontal",
              "paddingEnd": "md",
              "paddingStart": "md",
              "spacing": "lg",
              "type": "box",
            },
            {
              "backgroundColor": "#C7CADAAA",
              "borderWidth": "light",
              "contents": [],
              "layout": "horizontal",
              "offsetEnd": "5%",
              "offsetStart": "5%",
              "type": "box",
              "width": "95%",
            },
          ],
          "layout": "vertical",
          "paddingBottom": "md",
          "paddingTop": "lg",
          "spacing": "lg",
          "type": "box",
        },
        {
          "backgroundColor": "#C7CADA",
          "contents": [
            {
              "contents": [
                {
                  "alignItems": "center",
                  "contents": [
                    {
                      "size": "lg",
                      "text": "反",
                      "type": "text",
                    },
                    {
                      "align": "start",
                      "size": "lg",
                      "text": "思",
                      "type": "text",
                    },
                  ],
                  "flex": 1,
                  "justifyContent": "center",
                  "layout": "vertical",
                  "spacing": "xl",
                  "type": "box",
                },
                {
                  "contents": [
                    {
                      "size": "lg",
                      "text": "出埃及記 第八章",
                      "type": "text",
                      "weight": "bold",
                      "wrap": true,
                    },
                    {
                      "color": "#1C1C1D",
                      "flex": 9,
                      "size": "xs",
                      "text": "1. 你覺得神透過今天的經文對你說什麼呢?
2. 有什麼你可以做出的行動或改變呢?",
                      "type": "text",
                      "wrap": true,
                    },
                  ],
                  "flex": 11,
                  "layout": "vertical",
                  "paddingTop": "none",
                  "spacing": "md",
                  "type": "box",
                },
              ],
              "layout": "horizontal",
              "paddingEnd": "md",
              "paddingStart": "md",
              "spacing": "lg",
              "type": "box",
            },
            {
              "backgroundColor": "#FFFFFFAA",
              "borderWidth": "light",
              "contents": [],
              "layout": "horizontal",
              "offsetEnd": "5%",
              "offsetStart": "5%",
              "type": "box",
              "width": "95%",
            },
          ],
          "cornerRadius": "md",
          "layout": "vertical",
          "paddingBottom": "md",
          "paddingTop": "lg",
          "spacing": "lg",
          "type": "box",
        },
        {
          "backgroundColor": "#FFFFFF",
          "contents": [
            {
              "contents": [
                {
                  "alignItems": "center",
                  "contents": [
                    {
                      "size": "lg",
                      "text": "祈",
                      "type": "text",
                    },
                    {
                      "size": "lg",
                      "text": "求",
                      "type": "text",
                    },
                  ],
                  "flex": 1,
                  "justifyContent": "center",
                  "layout": "vertical",
                  "spacing": "xl",
                  "type": "box",
                },
                {
                  "contents": [
                    {
                      "color": "#1C1C1D",
                      "size": "xs",
                      "text": "”神啊！我將我的，___ ， ___ ， ___ 交給祢，我相信祢在這些事上掌權。",
                      "type": "text",
                      "wrap": true,
                    },
                    {
                      "contents": [
                        {
                          "contents": [
                            {
                              "borderColor": "#E2E2E299",
                              "borderWidth": "light",
                              "contents": [],
                              "layout": "vertical",
                              "type": "box",
                              "width": "1px",
                            },
                          ],
                          "flex": 1,
                          "layout": "horizontal",
                          "type": "box",
                        },
                        {
                          "color": "#1C1C1DAA",
                          "flex": 24,
                          "size": "xs",
                          "text": "『我們在天上的父：願人都尊你的名為聖。願你的國降臨；願你的旨意行在地上，如同行在天上。我們日用的飲食，今日賜給我們。免我們的債，如同我們免了人的債。不叫我們陷入試探；救我們脫離那惡者。因為國度、權柄、榮耀，全是你的，直到永遠。阿們！』”",
                          "type": "text",
                          "wrap": true,
                        },
                      ],
                      "layout": "horizontal",
                      "type": "box",
                    },
                  ],
                  "flex": 11,
                  "layout": "vertical",
                  "spacing": "md",
                  "type": "box",
                },
              ],
              "layout": "horizontal",
              "paddingEnd": "md",
              "paddingStart": "md",
              "spacing": "sm",
              "type": "box",
            },
            {
              "backgroundColor": "#C7CADAAA",
              "borderWidth": "light",
              "contents": [],
              "layout": "horizontal",
              "offsetEnd": "5%",
              "offsetStart": "5%",
              "type": "box",
              "width": "95%",
            },
          ],
          "layout": "vertical",
          "paddingBottom": "md",
          "paddingTop": "lg",
          "spacing": "lg",
          "type": "box",
        },
      ],
      "layout": "vertical",
      "type": "box",
    },
    "header": {
      "backgroundColor": "#FFCC32",
      "contents": [
        {
          "alignItems": "center",
          "contents": [
            {
              "align": "end",
              "color": "#1D292E",
              "flex": 2,
              "size": "xl",
              "text": "好好靈修",
              "type": "text",
              "weight": "bold",
            },
            {
              "align": "center",
              "size": "md",
              "text": "X",
              "type": "text",
            },
            {
              "contents": [
                {
                  "text": "DAILY",
                  "type": "text",
                  "weight": "bold",
                },
                {
                  "text": "DEVOTION",
                  "type": "text",
                  "weight": "bold",
                },
              ],
              "flex": 5,
              "layout": "vertical",
              "type": "box",
            },
          ],
          "layout": "horizontal",
          "type": "box",
        },
      ],
      "layout": "horizontal",
      "paddingAll": "20px",
      "spacing": "md",
      "type": "box",
    },
    "size": "giga",
    "type": "bubble",
  },
  "type": "flex",
}
`;
